### PR TITLE
DCMAW-8965: issue a service token

### PR DESCRIPTION
**Pull request by ahickmann-gds**
​DCMAW-8965

### What changed

**Token Lambda** 
- Add functionality to generate a service token (**UNENCRYPTED FOR NOW - ENCRYPTION TO BE DONE IN A SEPARATE PR**)
- Add new environment variables (`MOCK_STS_BASE_URL`,  `KEY_STORAGE_BUCKET_NAME`, `ASYNC_BACKEND_BASE_URL`)
- Fix Lambda S3 permissions

**Other changes** 
- Update `sts-mock` README instructions
- Add environment name as argument to `publish_keys_to_s3.sh` script
- Fix `jwksHelperScript.mjs` so that the private key JWK contains the key ID (`kid`)

Deployed to the `dev` environment and tested by sending a POST request to the `/token` endpoint to get a service token. The service token signature is then verified with the public key that has been published to the` /.well-known/jwks.json` endpoint.

**Service token**
Screenshot 2024-09-19 at 12 26 03

**Public key**
Screenshot 2024-09-19 at 12 27 52

**Decoded service token & signature verified using the public key**
Screenshot 2024-09-19 at 12 28 18

**Bad Request**
Screenshot 2024-09-19 at 14 40 45

Screenshot 2024-09-19 at 14 40 52

Screenshot 2024-09-19 at 14 40 58

**Server Error**
Screenshot 2024-09-19 at 14 48 34

## Checklists

- [x] There is a ticket raised for this PR that is present in the branch name
- [x] No PII data logged. [See guidance here](https://govukverify.atlassian.net/wiki/spaces/DCMAW/pages/3502407722/PII+Logging+Considerations)
- [ ] Demo to a BA, TA, and the team.
- [x] Update [README](./blob/main/README.md) with any new instructions or tasks

### CODE REVIEW

File: sts-mock/README.md [hunk 1]
```diff
@@ -45,7 +45,7 @@ npm run lint
 To manually deploy changes made to the stack (i.e. resources or source code) to the `dev` AWS account, run the following command after assuming your credentials:
 ```shell
 sam build --parallel --cached --beta-features
-sam deploy --guided
+sam deploy --guided --capabilities CAPABILITY_NAMED_IAM
```
Comment (jmooney-dd) [inline sts-mock/README.md:48]: Are there any specific values a dev should enter when using the guided flag? If so would it be worth adding them to the `README`?

File: sts-mock/jwks-helper-script/publish_keys_to_s3.sh [hunk 2]
```diff
@@ -9,7 +9,7 @@ if [ $? -gt 0 ]; then
   exit 1
 fi
 
-if [ $# != 1 ]; then
+if [ $# != 2 ]; then
   echo "Incorrect number of parameters supplied. Supply the stack name." 1>&2
```
Comment (jmooney-dd) [inline sts-mock/jwks-helper-script/publish_keys_to_s3.sh:13]: As there's 2 arguments now required should we update this line to ask them to supply the stack name and the env name?

File: sts-mock/src/token/serviceTokenGenerator/serviceTokenGenerator.ts [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+      return successResult(await response.Body!.transformToString());
+    } catch {
+      return errorResult({
+        errorMessage: "Unable to fetch file from S3",
+        errorCategory: "SERVER_ERROR",
+      });
+    }
+  }
+
+  private async formatKeyAsKeyLike(
```
Comment (jmooney-dd) [inline sts-mock/src/token/serviceTokenGenerator/serviceTokenGenerator.ts:76]: Is it possible to rename this to something specific? I don't have a good suggestion as I've not looked at what the importJWK func does yet

Comment (jmooney-dd) [reply sts-mock/src/token/serviceTokenGenerator/serviceTokenGenerator.ts:76]: Or is key like an actual term that's commonly used? If so then I wouldn't see this as a problem

Comment (jmooney-dd) [reply sts-mock/src/token/serviceTokenGenerator/serviceTokenGenerator.ts:76]: Ah I can see it's a type from Jose. Maybe this is fine then!

File: sts-mock/jwks-helper-script/publish_keys_to_s3.sh (hunk 2 above)
Comment (ahickmann-gds) [reply sts-mock/jwks-helper-script/publish_keys_to_s3.sh:13]: Yes, I will do that, thank you for pointing that out.

File: sts-mock/README.md (hunk 1 above)
Comment (ahickmann-gds) [reply sts-mock/README.md:48]: Can do, yes - I will add those too. I suppose those values are usually more or less the same and we get used to entering them after doing it once or twice, but it's always helpful for someone doing it the first time!

File: sts-mock/README.md
Comment (jhumbert-dd) [inline sts-mock/README.md:1]: Not part of the changes in this PR but should [line 16](https://github.com/govuk-one-login/mobile-id-check-async/pull/154/files#diff-879e1b4afbc7bd75ae1a5274662450c942f7127ff03dd1a8eb661f4d1464b577R16:~:text=npm%20run-,install,-%60%60%60) be `npm install`? I don't think the install works with `run`

Comment (ahickmann-gds) [reply sts-mock/README.md:1]: Yes! Thank you!

File: sts-mock/README.md [hunk 4]
```diff
@@ -44,13 +44,44 @@ npm run lint
 ### Manual Deployment
 To manually deploy changes made to the stack (i.e. resources or source code) to the `dev` AWS account, run the following command after assuming your credentials:
 ```shell
-sam build --parallel --cached --beta-features
```
Comment (ahickmann-gds) [inline sts-mock/README.md:47]: I don't think `--parallel` is needed

File: sts-mock/jwks-helper-script/jwksHelperScript.mjs [hunk 5]
```diff
@@ -9,18 +8,22 @@ const keyPair = await generateKeyPair("ES256", {
...
 
 const publicKey = await exportJWK(keyPair.publicKey);
-const jwks = {
+
+const publicKeyJwks = {
     keys: [
         {
             ...publicKey,
-            kid: randomUUID(),
+            kid: thumbprint,
```
Comment (ahickmann-gds) [inline sts-mock/jwks-helper-script/jwksHelperScript.mjs:21]: Using the actual key ID instead of a random UUID. I should have done this in the first place!

File: sts-mock/src/token/tokenHandler.ts [hunk 6]
```diff
@@ -1,16 +1,75 @@
...
+import {
+  APIGatewayProxyEvent,
+  APIGatewayProxyResult,
+  Context,
+} from "aws-lambda";
 import { Dependencies, dependencies } from "./handlerDependencies";
+import { ConfigService } from "./configService/configService";
+import { JWTPayload } from "jose";
+
+const SERVICE_TOKEN_TTL_IN_SECS = 180;
```
Comment (ahickmann-gds) [inline sts-mock/src/token/tokenHandler.ts:10]: Putting these constants up here to make them obvious

File: sts-mock/src/token/validateServiceTokenRequest/validateServiceTokenRequest.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
+import { errorResult, Result, successResult } from "../../utils/result";
```
Comment (ahickmann-gds) [inline sts-mock/src/token/validateServiceTokenRequest/validateServiceTokenRequest.ts:1]: I think we need to have a conversation about how much request validation we want to do/is necessary in this mock. This can be enhanced in a separate PR.

File: sts-mock/src/token/tokenHandler.ts [hunk 8]
```diff
@@ -22,11 +81,51 @@ export async function lambdaHandlerConstructor(
...
+
+function getServiceTokenPayload(
+  issuer: string,
+  tokenExpiry: number,
+  sub: string,
+  scope: string,
+): JWTPayload {
+  const nowInSeconds = Math.floor(Date.now() / 1000);
+  return {
+    aud: "TBC",
```
Comment (ahickmann-gds) [inline sts-mock/src/token/tokenHandler.ts:124]: What should the value of the audience claim be?

File: sts-mock/template.yaml [hunk 9]
```diff
@@ -145,6 +145,11 @@ Resources:
             Method: post
             RestApiId: !Ref StsMockApi
       Role: !GetAtt TokenFunctionLambdaRole.Arn
+      Environment:
+        Variables:
+          # MOCK_STS_BASE_URL to be updated once available
+          MOCK_STS_BASE_URL: MockStsBaseUrl
```
Comment (ahickmann-gds) [inline sts-mock/template.yaml:151]: Do we have a value for MOCK_STS_BASE_URL?

**Comment by ahickmann-gds** [comment]
**To do in the next PR:**
- Encrypt service token
- DNS
- API tests

File: sts-mock/jwks-helper-script/jwksHelperScript.mjs
Comment (sandymay-dd) [inline sts-mock/jwks-helper-script/jwksHelperScript.mjs:1]: Whilst we're here - please may we update this file name to state its purpose?

File: sts-mock/openapi/sts-mock-spec.yaml [hunk 10]
```diff
@@ -96,7 +96,7 @@ paths:
     post:
       operationId: postToken
       summary: Obtain a service token.
-      description: Exchange a Mock STS-issued access token for a service token.
+      description: Exchange an access token for a mock service token.
```
Comment (sandymay-dd) [inline sts-mock/openapi/sts-mock-spec.yaml:99]: Should this say exchange a mock subject token for a mock service token?

File: sts-mock/src/testUtils/getMockSigningKey.ts [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+    kty: "EC",
+    x: "X5Sk_TNlp2cZ_mUQ7p_Gqth7IVkLGaMFfUH-e6pzUC8",
+    y: "Iw5VG-5Sb-c954Xqanov47QhQVLEw3H9BdGNwR9RvOw",
+    crv: "P-256",
+    d: "eswmWYW980vOl7x0MZBqzAkxN215eogA0fFJCu6umfw",
+    kid: "iyVpkshZ0QKq5ORWz7mc76x0dAKUp4RS113tiHACjpQ",
+  };
+  const privateKey = createPrivateKey({ key: jwk, format: "jwk" });
+  const keyId = jwk.kid;
+  return { signingKey: privateKey, keyId };
```
Comment (sandymay-dd) [inline sts-mock/src/testUtils/getMockSigningKey.ts:15]:
```suggestion
  const signingKey = createPrivateKey({ key: jwk, format: "jwk" });
  const keyId = jwk.kid;
  return { signingKey, keyId };
```

Thoughts? Just to avoid using two different terms for the same thing

Comment (sandymay-dd) [reply sts-mock/src/testUtils/getMockSigningKey.ts:15]: Or rename the interface to expect `privateKey` if you prefer that variable name

File: sts-mock/src/testUtils/getMockSigningKey.ts
Comment (sandymay-dd) [inline sts-mock/src/testUtils/getMockSigningKey.ts:1]: Please may we double check whether these are excluded in our sonar properties file

File: sts-mock/src/token/keyRetriever/keyRetriever.ts [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+      Bucket: bucketName,
+      Key: fileName,
+    };
+
+    let key;
+    try {
+      const response = await this.s3Client.send(
+        new GetObjectCommand(commandInput),
+      );
+      key = await response.Body!.transformToString();
```
Comment (sandymay-dd) [inline sts-mock/src/token/keyRetriever/keyRetriever.ts:45]: Should we look at handling the scenario where the body doesn't come through? I'm not too keen on using the `!`

File: sts-mock/src/token/keyRetriever/keyRetriever.ts [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+
+  private formatKeyAsKeyObject(key: string): Result<SigningKey> {
+    try {
+      const jwk = this.parseAsJwk(key);
+
+      // Convert from JWK to a runtime-specific key representation (KeyObject).
+      const privateKey = createPrivateKey({ key: jwk, format: "jwk" });
+      const keyId = jwk.kid;
+
+      return successResult({ signingKey: privateKey, keyId });
```
Comment (sandymay-dd) [inline sts-mock/src/token/keyRetriever/keyRetriever.ts:64]: Similar here - feels strange we have two terms for the same thing (signingKey + privateKey)

File: sts-mock/src/token/keyRetriever/tests/keyRetriever.test.ts [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+import { KeyRetriever, SigningKey } from "../keyRetriever";
+
+const mockS3Client = mockClient(S3Client);
+const mockS3Response = (content: string) => {
+  return {
+    transformToString: async () => content,
+  };
+};
+
+describe("Key Retriever", () => {
```
Comment (sandymay-dd) [inline sts-mock/src/token/keyRetriever/tests/keyRetriever.test.ts:16]: reminder - add a test for the scenario where the body is undefined in the S3 response :)

File: sts-mock/src/token/tokenHandler.test.ts [hunk 15]
```diff
@@ -1,41 +1,184 @@
...
+        expect(result.statusCode).toStrictEqual(500);
+        expect(result.body).toStrictEqual(
+          '{"error":"server_error","error_description":"Server Error"}',
+        );
+      });
+    });
+  });
+
+  describe("Validate Service Token Request", () => {
+    describe("Given the request body is invalid (e.g. undefined)", () => {
```
Comment (sandymay-dd) [inline sts-mock/src/token/tokenHandler.test.ts:71]: I think it would be good to look at covering the different permutations here please

e.g. Given there is no scope, Given there is no subject token etc.

File: sts-mock/src/token/tokenHandler.test.ts [hunk 16]
```diff
@@ -1,41 +1,184 @@
...
+        expect(result.statusCode).toStrictEqual(500);
+        expect(result.body).toStrictEqual(
+          '{"error":"server_error","error_description":"Server Error"}',
+        );
+      });
+    });
+  });
+
+  describe("Token Signer", () => {
+    describe("Given an error happens trying to crate a signed token", () => {
```
Comment (sandymay-dd) [inline sts-mock/src/token/tokenHandler.test.ts:127]:
```suggestion
    describe("Given an error happens trying to create a signed token", () => {
```

File: sts-mock/src/token/tokenHandler.test.ts [hunk 17]
```diff
@@ -1,41 +1,184 @@
...
+          headers: {
+            "Content-Type": "application/json",
+            "Cache-Control": "no-store",
+            Pragma: "no-cache",
+          },
+          statusCode: 200,
+          body: JSON.stringify({
+            access_token: "mockServiceToken",
+            token_type: "Bearer",
+            expires_in: 180,
```
Comment (sandymay-dd) [inline sts-mock/src/token/tokenHandler.test.ts:179]: Is this expiry value corect?

File: sts-mock/src/token/validateServiceTokenRequest/validateServiceTokenRequest.ts [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+  const subjectToken = searchParams.get("subject_token");
+  if (!subjectToken) {
+    return errorResult({
+      errorMessage: "Missing subject_token",
+      errorCategory: "CLIENT_ERROR",
+    });
+  }
+
+  const scope = searchParams.get("scope");
+  if (!scope) {
```
Comment (sandymay-dd) [inline sts-mock/src/token/validateServiceTokenRequest/validateServiceTokenRequest.ts:32]: I'm wondering if we should also check the scope value. This would guard against the possibility of us accidentally changing it in one of our lambdas?

File: sts-mock/template.yaml [hunk 19]
```diff
@@ -145,6 +145,11 @@ Resources:
             Method: post
             RestApiId: !Ref StsMockApi
       Role: !GetAtt TokenFunctionLambdaRole.Arn
+      Environment:
+        Variables:
+          # MOCK_STS_BASE_URL to be updated once available
```
Comment (sandymay-dd) [inline sts-mock/template.yaml:150]: I'd suggest also passing through the AUD value here - it'll need to be dynamic per environment. 

The value to answer your question elsewhere will be:

https://mob-async-backend.review-b-async.<ENVIRONMENT>.account.gov.uk

Comment (ahickmann-gds) [reply sts-mock/template.yaml:150]: Ahhh ok, I didn't realise it will change per environment, but I should have 🤦 - thanks!

File: sts-mock/src/token/tokenHandler.test.ts (hunk 17 above)
Comment (ahickmann-gds) [reply sts-mock/src/token/tokenHandler.test.ts:179]: As per https://govukverify.atlassian.net/wiki/spaces/DCMAW/pages/3844964353/Strategic+App+App+calls+a+protected+service#Example-Response and https://github.com/govuk-one-login/sts-back/blob/cc84af37f4354458a1b1ac12c8c381df5710ad2b/backend-api/template.yaml#L2912, it is correct, unless I'm not looking in the right place

File: sts-mock/jwks-helper-script/jwksHelperScript.mjs
Comment (ahickmann-gds) [reply sts-mock/jwks-helper-script/jwksHelperScript.mjs:1]: Done - it's called `generateKeyPair` now, but I can rename it. I've added some comments to this file and also added a comment to the bash script.

File: sts-mock/template.yaml (hunk 9 above)
Comment (ahickmann-gds) [reply sts-mock/template.yaml:151]: This will be addressed in the next PR (DNS)

File: sts-mock/src/token/tokenHandler.test.ts (hunk 15 above)
Comment (ahickmann-gds) [reply sts-mock/src/token/tokenHandler.test.ts:70]: Yes, I had done that originally, but because I added those scenarios to validateServiceTokenRequest.test.ts, I didn't want to repeat them here, it didn't seem necessary because the logic in the handler is the same? What do you think?

File: sts-mock/src/testUtils/getMockSigningKey.ts (hunk 11 above)
Comment (ahickmann-gds) [reply sts-mock/src/testUtils/getMockSigningKey.ts]: Good idea, done!

File: sts-mock/openapi/sts-mock-spec.yaml (hunk 10 above)
Comment (ahickmann-gds) [reply sts-mock/openapi/sts-mock-spec.yaml:99]: It's referred to as STS access token in the documentation. Can we have this instead?
`description: Exchange an STS access token (subject token) for a mock service token.`

File: sts-mock/src/token/keyRetriever/keyRetriever.ts (hunk 13 above)
Comment (ahickmann-gds) [reply sts-mock/src/token/keyRetriever/keyRetriever.ts]: Done!

File: sts-mock/template.yaml (hunk 19 above)
Comment (ahickmann-gds) [reply sts-mock/template.yaml:150]: Done - `ASYNC_BACKEND_BASE_URL`

File: sts-mock/src/token/tokenHandler.ts (hunk 8 above)
Comment (ahickmann-gds) [reply sts-mock/src/token/tokenHandler.ts]: Dealt with below.

File: sts-mock/src/testUtils/getMockSigningKey.ts
Comment (ahickmann-gds) [reply sts-mock/src/testUtils/getMockSigningKey.ts:1]: It seems to be as it's in `testUtils`

`sonar.exclusions=**/*.test.ts,**/*.test.js,**/testUtils/**,**/tests/mocks.ts, **/tests**`

File: sts-mock/openapi/sts-mock-spec.yaml (hunk 10 above)
Comment (sandymay-dd) [reply sts-mock/openapi/sts-mock-spec.yaml]: Absolutely - just wanted to get subject token _somewhere_ in there

File: sts-mock/src/testUtils/getMockSigningKey.ts
Comment (sandymay-dd) [reply sts-mock/src/testUtils/getMockSigningKey.ts:1]: perfect - thank you!

File: sts-mock/src/token/keyRetriever/keyRetriever.ts (hunk 12 above)
Comment (sandymay-dd) [reply sts-mock/src/token/keyRetriever/keyRetriever.ts]: I think let's put our best foot forward unless we need to consciously make a decision to reduce quality to reduce engineering effort. It's a good question to ask and one that we could debate :D

File: sts-mock/src/token/tokenHandler.test.ts (hunk 15 above)
Comment (sandymay-dd) [reply sts-mock/src/token/tokenHandler.test.ts:71]: No need to duplicate - apologies!

Comment (sandymay-dd) [reply sts-mock/src/token/tokenHandler.test.ts:71]: I'm wondering here: if we aren't mocking this validation functionality in the lambda - do you think the unit tests would be better suited to be in the lambda handler unit tests?

By moving these unit tests to the lambda unit test file, we can refactor the validateServiceTokenRequest.ts file as much as we want without changing a single unit test.

File: sts-mock/src/token/keyRetriever/keyRetriever.ts (hunk 12 above)
Comment (ahickmann-gds) [reply sts-mock/src/token/keyRetriever/keyRetriever.ts]: After doing some research, it seems that `transformToString()` will throw an error if `Body` is falsy, but as this line of code is in the catch block, that error will be caught. However, I have now added an if statement to check if `Body` is falsy and return a different error message if it is. Please have a look. 🥺

File: sts-mock/src/token/keyRetriever/tests/keyRetriever.test.ts (hunk 14 above)
Comment (ahickmann-gds) [reply sts-mock/src/token/keyRetriever/tests/keyRetriever.test.ts:16]: Done - see line 46 of this file

File: sts-mock/src/token/tokenHandler.test.ts (hunk 16 above)
Comment (ahickmann-gds) [reply sts-mock/src/token/tokenHandler.test.ts]: Done :)

File: sts-mock/src/token/tokenHandler.test.ts (hunk 15 above)
Comment (ahickmann-gds) [reply sts-mock/src/token/tokenHandler.test.ts:71]: I think that's a good idea. I was not 100% sure about where they should go. I will move them over now...

Comment (ahickmann-gds) [reply sts-mock/src/token/tokenHandler.test.ts]: Done

File: sts-mock/src/token/validateServiceTokenRequest/validateServiceTokenRequest.ts (hunk 7 above)
Comment (ahickmann-gds) [reply sts-mock/src/token/validateServiceTokenRequest/validateServiceTokenRequest.ts:1]: Discussed below.

**Review by sandymay-dd — DISMISSED** [review]

File: sts-mock/src/token/keyRetriever/keyRetriever.ts [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+
+    const { Body } = response;
+    if (!Body) {
+      return errorResult({
+        errorMessage: "Empty object retrieved from S3",
+        errorCategory: "SERVER_ERROR",
+      });
+    }
+
+    const signingKey = await Body.transformToString();
```
[the remark this answers is not in the document]
Comment (ahickmann-gds) [reply sts-mock/src/token/keyRetriever/keyRetriever.ts]: Changed!

**Review by sandymay-dd — APPROVED** [review]
